File: android/src/main/java/com/authsignal/react/AuthsignalPushModule.java
package com.authsignal.react;

import android.util.Log;

import androidx.annotation.NonNull;

import com.authsignal.push.AuthsignalPush;
import com.authsignal.push.models.PushChallenge;
import com.authsignal.push.models.PushCredential;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableMap;

import java.util.HashMap;
import java.util.Map;

public class AuthsignalPushModule extends ReactContextBaseJavaModule {
  private final ReactApplicationContext reactContext;

  private AuthsignalPush authsignalPush;

  private final String TAG = "AuthsignalPasskeyModule";
  private final String INIT_WARNING = "AuthsignalPasskeyModule is not initialized.";

  public AuthsignalPushModule(ReactApplicationContext reactContext) {
    super(reactContext);
    this.reactContext = reactContext;
  }

  @Override
  public Map<String, Object> getConstants() {
    final Map<String, Object> constants = new HashMap<>();
    constants.put("bundleIdentifier", reactContext.getApplicationInfo().packageName);
    return constants;
  }

  @NonNull
  @Override
  public String getName() {
    return "AuthsignalPushModule";
  }

  @ReactMethod
  public void initialize(String tenantID, String baseURL, Promise promise) {
    authsignalPush = new AuthsignalPush(tenantID, baseURL);

    promise.resolve(null);
  }

  @ReactMethod
  public void getCredential(Promise promise) {
    if (authsignalPush != null) {
      authsignalPush
        .getCredentialAsync()
        .thenAcceptAsync((response) -> {
          if (response.getError() != null) {
            promise.reject("get_credential_error", response.getError());
          } else {
            PushCredential credential = response.getData();
            WritableMap map = Arguments.createMap();
            map.putString("credentialId", credential.getCredentialId());
            map.putString("createdAt", credential.getCreatedAt());
            map.putString("lastAuthenticatedAt", credential.getLastAuthenticatedAt());
            promise.resolve(map);
          }
        });
    } else {
      Log.w(TAG, INIT_WARNING);

      promise.resolve(null);
    }
  }

  @ReactMethod
  public void addCredential(
    String token,
    Promise promise) {
    if (authsignalPush != null) {
      authsignalPush
        .addCredentialAsync(token, null)
        .thenAcceptAsync(response -> {
          String errorCode = response.getErrorType() != null ?
            response.getErrorType() :
            "add_credential_error";

          if (response.getError() != null) {
            promise.reject(errorCode, response.getError());
          } else {
            promise.resolve(response.getData());
          }
        });
    } else {
      Log.w(TAG, INIT_WARNING);

      promise.resolve(false);
    }
  }

  @ReactMethod
  public void removeCredential(Promise promise) {
    if (authsignalPush != null) {
      authsignalPush
        .removeCredentialAsync()
        .thenAcceptAsync(response -> {
          if (response.getError() != null) {
            String errorCode = response.getErrorType() != null ?
              response.getErrorType() :
              "remove_credential_error";

            promise.reject(errorCode, response.getError());
          } else {
            promise.resolve(response.getData());
          }
        });
    } else {
      Log.w(TAG, INIT_WARNING);

      promise.resolve(false);
    }
  }

  @ReactMethod
  public void getChallenge(Promise promise) {
    if (authsignalPush != null) {
      authsignalPush
        .getChallengeAsync()
        .thenAcceptAsync(response -> {
          if (response.getError() != null) {
            String errorCode = response.getErrorType() != null ?
              response.getErrorType() :
              "get_challenge_error";

            promise.reject(errorCode, response.getError());
          } else {
            PushChallenge challenge = response.getData();

            if (challenge == null) {
              promise.resolve(null);
            } else {
              WritableMap map = Arguments.createMap();
              map.putString("challengeId", challenge.getChallengeId());
              map.putString("actionCode", challenge.getActionCode());
              map.putString("idempotencyKey", challenge.getIdempotencyKey());
              map.putString("ipAddress", challenge.getIpAddress());
              map.putString("deviceId", challenge.getDeviceId());
              map.putString("userAgent", challenge.getUserAgent());
              promise.resolve(map);
            }
          }
        });
    } else {
      Log.w(TAG, INIT_WARNING);

      promise.resolve(null);
    }
  }

  @ReactMethod
  public void updateChallenge(
    String challengeId,
    Boolean approved,
    String verificationCode,
    Promise promise
  ) {
    if (authsignalPush != null) {
      authsignalPush
        .updateChallengeAsync(challengeId, approved, verificationCode)
        .thenAcceptAsync(response -> {
          if (response.getError() != null) {
            String errorCode = response.getErrorType() != null ?
              response.getErrorType() :
              "update_challenge_error";

            promise.reject(errorCode, response.getError());
          } else {
            promise.resolve(response.getData());
          }
        });
    } else {
      Log.w(TAG, INIT_WARNING);

      promise.resolve(false);
    }
  }
}

File: android/src/main/java/com/authsignal/react/AuthsignalPushModule.kt
package com.authsignal.react

import android.util.Log
import com.authsignal.push.AuthsignalPush
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.module.annotations.ReactModule
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch

@ReactModule(name = AuthsignalPushModule.NAME)
class AuthsignalPushModule(private val reactContext: ReactApplicationContext) :
  NativeAuthsignalPushModuleSpec(reactContext) {
  private val coroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
  private var authsignal: AuthsignalPush? = null
  private var defaultError = "unexpected_error"

  @ReactMethod
  override fun initialize(tenantID: String, baseURL: String, promise: Promise) {
    authsignal = AuthsignalPush(tenantID, baseURL)

    promise.resolve(null)
  }

  @ReactMethod
  override fun getCredential(promise: Promise) {
    launch(promise) {
      val response = it.getCredential()

      if (response.error != null) {
        val errorCode = response.errorCode ?: defaultError

        promise.reject(errorCode, response.error)
      } else if (response.data != null) {
        val credential = response.data
        val map = Arguments.createMap()
        map.putString("credentialId", credential!!.credentialId)
        map.putString("createdAt", credential.createdAt)
        map.putString("userId", credential.userId)
        map.putString("lastAuthenticatedAt", credential.lastAuthenticatedAt)
        promise.resolve(map)
      } else {
        promise.resolve(null)
      }
    }
  }

  @ReactMethod
  override fun addCredential(
    token: String?,
    _requireUserAuthentication: Boolean,
    _keychainAccess: String?,
    performAttestation: Boolean,
    promise: Promise
  ) {
    launch(promise) {
      val response = it.addCredential(token, null, performAttestation = performAttestation)

      if (response.error != null) {
        val errorCode = response.errorCode ?: defaultError

        promise.reject(errorCode, response.error)
      } else {
        val credential = response.data
        val map = Arguments.createMap()
        map.putString("credentialId", credential!!.credentialId)
        map.putString("createdAt", credential.createdAt)
        map.putString("userId", credential.userId)
        map.putString("lastAuthenticatedAt", credential.lastAuthenticatedAt)
        promise.resolve(map)
      }
    }
  }

  @ReactMethod
  override fun removeCredential(promise: Promise) {
    launch(promise) {
      val response = it.removeCredential()

      if (response.error != null) {
        val errorCode = response.errorCode ?: defaultError

        promise.reject(errorCode, response.error)
      } else {
        promise.resolve(response.data)
      }
    }
  }

  @ReactMethod
  override fun getChallenge(promise: Promise) {
    launch(promise) {
      val response = it.getChallenge()

      if (response.error != null) {
        val errorCode = response.errorCode ?: defaultError

        promise.reject(errorCode, response.error)
      } else {
        val challenge = response.data

        if (challenge == null) {
          promise.resolve(null)
        } else {
          val map = Arguments.createMap()
          map.putString("challengeId", challenge.challengeId)
          map.putString("actionCode", challenge.actionCode)
          map.putString("idempotencyKey", challenge.idempotencyKey)
          map.putString("ipAddress", challenge.ipAddress)
          map.putString("deviceId", challenge.deviceId)
          map.putString("userAgent", challenge.userAgent)
          promise.resolve(map)
        }
      }
    }
  }

  @ReactMethod
  override fun updateChallenge(
    challengeId: String,
    approved: Boolean,
    verificationCode: String?,
    promise: Promise
  ) {
    launch(promise) {
      val response = it.updateChallenge(challengeId, approved, verificationCode)

      if (response.error != null) {
        val errorCode = response.errorCode ?: defaultError

        promise.reject(errorCode, response.error)
      } else {
        promise.resolve(response.data)
      }
    }
  }

  private fun launch(promise: Promise, fn: suspend (client: AuthsignalPush) -> Unit) {
    coroutineScope.launch {
      authsignal?.let {
        fn(it)
      }  ?: run {
        Log.w("init_error", "AuthsignalPushModule is not initialized.")

        promise.resolve(null)
      }
    }
  }

  companion object {
    const val NAME = "AuthsignalPushModule"
  }
}
